### ISSUE TITLE
The old error "can't add a new key into hash during iteration"

### ISSUE THREAD

**Issue by sasharevzin** [newrelic/newrelic-ruby-agent#933]
The same [issue](https://github.com/newrelic/newrelic-ruby-agent/issues/573) happens and wasn't resolved yet.

## Description
When an exception is triggered and the puma runs in multiple threads and workers it throws `can't add a new key into hash during iteration` error. Yes, same `request.rb:79:in set_header'` as it was in [this](https://github.com/newrelic/newrelic-ruby-agent/issues/573) issue
```
Error during failsafe response: can't add a new key into hash during iteration
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/request.rb:79:in `set_header'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/actionpack-5.2.6/lib/action_dispatch/middleware/show_exceptions.rb:48:in `render_exception'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/actionpack-5.2.6/lib/action_dispatch/middleware/show_exceptions.rb:36:in `rescue in call'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/actionpack-5.2.6/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/newrelic_rpm-8.4.0/lib/new_relic/agent/instrumentation/middleware_tracing.rb:100:in `call'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/railties-5.2.6/lib/rails/rack/logger.rb:38:in `call_app'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/railties-5.2.6/lib/rails/rack/logger.rb:26:in `block in call'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/activesupport-5.2.6/lib/active_support/tagged_logging.rb:71:in `block in tagged'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/activesupport-5.2.6/lib/active_support/tagged_logging.rb:28:in `tagged'
  .rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/activesupport-5.2.6/lib/active_support/tagged_logging.rb:71:in `tagged'
```

## Expected Behavior
The error shouldn't happen

## Steps to Reproduce
I created [a brand new Rails app](https://github.com/sasharevzin/newrelic_app) (added master.key as well) and can 100% time reproduce the issue.
1. Start with a production environment like `RAILS_ENV=production bundle exec rails s`
2. Monitor logs by running `tail -f log/production.log`
3. Go to http://localhost:3000
4. You'll get the page `we are sorry...`
5. Now refresh the page multiple times by hitting `COMMAND + R` in the Chrome browser

## Your Environment
It's a brand new rails app with only `newrelic_rpm` gem on it
```
ruby '2.6.8'
rails', '~> 5.2.4'
newrelic_rpm (8.4.0)
```

**Comment by kaylareopelle** [comment]
Hi @sasharevzin! Thank you for your comment on the old issue yesterday and for creating a new issue today. I'm sorry to hear this issue persists. I'll take a look at this issue today and reach out with any follow-up questions.

**Comment by kaylareopelle** [comment]
Hi @sasharevzin. Thank you for the repository that reproduces the issue! It's been a huge help. Our team was able to replicate the behavior are working on a solution.

**Comment by tannalynn** [comment]
Hello @sasharevzin,
Thank you for bringing our attention back to this issue and letting us know it's still going on. I believe I have identified the source of the error and have created pr #942 which should resolve the issue and prevent it from happening. With this fix, I was no longer able to cause the error to occur using the reproduction you provided. Thank you so much for providing that reproduction, I found it very valuable in my investigation. It is very much appreciated. 

Since this error has continued to occur despite previous fix attempts, it would be great if you are able to test this branch out and confirm for us if the issue appears to be resolved for you. You can do this by using this in your gemfile
```
gem 'newrelic_rpm', :git => 'https://github.com/newrelic/newrelic-ruby-agent.git',  :branch => '933_cant_add_new_key_hash'
```
Thank you!

**Comment by sharkey11** [comment]
@sasharevzin 🐐

**Comment by sasharevzin** [comment]
@tannalynn I can confirm that it works and the error is no longer happening. 🥳

**Comment by tannalynn** [comment]
@sasharevzin 
That's awesome! I'm so glad to hear that. 🎉 
Thanks for getting back to me about it so quick!  This fix will be included in our next release (8.5.0).
